kubeadm: use bind-address instead of address to configure the http probes

aojea opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug

/kind cleanup

> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

kubeadm removed the deprecated "--address" flag for controller-manager
and scheduler in favor of "--bind-address" in #71973

We should use bind-address to configure the manifest probe addresses.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

Found this in https://github.com/kubernetes-sigs/kind/pull/1195

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kubeadm: use bind-address option to configure the kube-controller-manager and kube-scheduler http probes
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aojea committed:
kubeadm: use bind-address instead of address

kubeadm removed the deprecated "--address" flag for controller-manager
and scheduler in favor of "--bind-address"

We should use bind-address to configure the manifest probe addresses.

aojea commented:
/assign @neolit123 @ereslibre 
/cc @BenTheElder 
/sig cluster-lifecycle

liggitt commented on cmd/kubeadm/app/util/staticpod/utils.go:
```diff
@@ -252,15 +252,15 @@ func GetAPIServerProbeAddress(endpoint *kubeadmapi.APIEndpoint) string {
 // GetControllerManagerProbeAddress returns the kubernetes controller manager probe address
 func GetControllerManagerProbeAddress(cfg *kubeadmapi.ClusterConfiguration) string {
-	if addr, exists := cfg.ControllerManager.ExtraArgs[kubeControllerManagerAddressArg]; exists {
+	if addr, exists := cfg.ControllerManager.ExtraArgs[kubeControllerManagerBindAddressArg]; exists {
```
will a probe address of `0.0.0.0` work?

aojea replied:
good catch 
```
--- FAIL: TestGetControllerManagerProbeAddress (0.00s)
    --- FAIL: TestGetControllerManagerProbeAddress/setting_controller_manager_extra_ipv6_address_arg_to_something_acknowledges_it#01 (0.00s)
        /Users/aojea/work/src/k8s.io/kubernetes/cmd/kubeadm/app/util/staticpod/utils_test.go:141: Unexpected result from GetControllerManagerProbeAddress:
            	expected: 127.0.0.1
            	actual: 0.0.0.0
```

aojea replied:
I guess that we should default to localhost for unspecified addresses, right?
my question is, should we return `localhost` or try to be smart and resolve the ip family?

aojea replied:
@mauilion I remember that you told me that in linux is possible to use 0.0.0.0 as destination address, but do you think that will be safe to use it to configure the http probe?

neolit123 replied:
it feels to me that if `0.0.0.0` is passed as `--bind-address` it should be used for a probe too, transparently, without defaulting to a localhost or per family IP.

aojea replied:
I think that we should fall back to loopbacks addresses for the probes, it's the most secure and reliable IMHO

neolit123 replied:
alright.

aojea replied:
I mean, if it binds to 0.0.0.0 is going to listen in all interfaces and all ip addresses ... but I don't understand what IP is going to poll the probe, is deterministic the 0.0.0.0 or it just picks one randomly in the host?

liggitt replied:
> I mean, if it binds to 0.0.0.0 is going to listen in all interfaces and all ip addresses ... but I don't understand what IP is going to poll the probe, is deterministic the 0.0.0.0 or it just picks one randomly in the host?

I have the same question... is contacting 0.0.0.0 a normal thing to do?

neolit123 replied:
we mostly care for the probes to serve on a loopback IP, yet the 0.0.0.0 -> loopback IP transform becomes a undocumented behavior.

if someone complains they can customize their manifests.

liggitt replied:
does a probe against 0.0.0.0 actually work?

aojea replied:
_checking_ in the meantime net.Dial considers it valid :/
https://golang.org/src/net/dial.go

ereslibre replied:
One thing I'm worried about here is whether using 0.0.0.0 for the probe works now because of obscure reasons but will stop working in the future; I'm not sure if this is legal at all. Also, would this work on kubelet on Windows?

aojea replied:
`kind create cluster`
kubectl apply
```
apiVersion: v1
kind: Pod
metadata:
  labels:
    test: liveness
  name: liveness-http
spec:
  containers:
  - name: liveness
    image: k8s.gcr.io/liveness
    args:
    - /server
    livenessProbe:
      httpGet:
        host: 0.0.0.0
        path: /healthz
        port: 8080
        httpHeaders:
        - name: Custom-Header
          value: Awesome
      initialDelaySeconds: 3
      periodSeconds: 3
```

and the winner is 😄 

```
Events:
  Type     Reason     Age                 From                         Message
  ----     ------     ----                ----                         -------
  Normal   Scheduled  109s                default-scheduler            Successfully assigned default/liveness-http to kind-control-plane
  Normal   Pulled     73s (x3 over 100s)  kubelet, kind-control-plane  Successfully pulled image "k8s.gcr.io/liveness"
  Normal   Created    72s (x3 over 99s)   kubelet, kind-control-plane  Created container liveness
  Normal   Started    72s (x3 over 99s)   kubelet, kind-control-plane  Started container liveness
  Normal   Pulling    63s (x4 over 108s)  kubelet, kind-control-plane  Pulling image "k8s.gcr.io/liveness"
  Warning  Unhealthy  63s (x9 over 96s)   kubelet, kind-control-plane  Liveness probe failed: Get http://0.0.0.0:8080/healthz: dial tcp 0.0.0.0:8080: connect: connection refused
  Normal   Killing    63s (x3 over 90s)   kubelet, kind-control-plane  Container liveness failed liveness probe, will be restarted
```

aojea replied:
I've created another one without the host parameter and it works

the curious is that the url without host works
```
    Liveness:       http-get http://:8080/healthz delay=3s timeout=1s period=3s #success=1 #failure=3
```
but this doesn't
```
    Liveness:       http-get http://0.0.0.0:8080/healthz delay=3s timeout=1s period=3s #success=1 #failure=3
```

neolit123 replied:
ok, in that case our fallback here is valid.

neolit123 replied:
> the curious is that the url without host works

this could be defaulting that k8s probes are doing.

aojea replied:
using 0.0.0.0 will end in a broken environment hard to debug (it took me a while today to find out the problem) and is common to use those addresses, at least we use `::` in kind , I think that in case of having bind-address as 0.0.0.0 or :: we should choose one of the following options:
1. omit the host field in the probe
2. use a loopback address

aojea replied:
I'm leaning towards omitting the host field, at the end 0.0.0.0 and :: is kind of "no address" :)

ereslibre replied:
So looks like omitting the port is the best way to go, so we don't need ad-hoc logic for ipv4 and ipv6 here, right?

neolit123 replied:
kubeadm itself does block 0.0.0.0 at least as an api server advertise address: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/config/common.go#L121-L140
but not for custom bind-address (extra args) for the other components.

we also have to note that the api server probe uses the advertise address, while the probes for the other components use bind-address
https://github.com/kubernetes/kubernetes/blob/34ba1683b9d68284a481814d58673f999cff0520/cmd/kubeadm/app/util/staticpod/utils.go#L237-L251

so it seems to me a api server probe of 0.0.0.0 was already not possible.

ereslibre replied:
For the sake of completeness: https://tools.ietf.org/html/rfc5735#section-3.

thockin replied:
Commented on #86504

aojea commented:
/hold 
https://github.com/kubernetes/kubernetes/pull/86493#discussion_r360576622

aojea committed:
kubeadm: probe address for unspecified ips

kubeadm deploys the apiserver, controller-manager and the scheduler
using liveness probes.
The bind-address option is used to configure the probe address, in
case this is configured with an unspecified address, the probe
will fail. When using an unspecified address the probe host field is
left empty, otherwise the bind-address is used.

aojea commented:
/test pull-kubernetes-e2e-kind-ipv6

neolit123 commented:
/lgtm
/approve

ereslibre commented:
Thank you @aojea.

/lgtm

Feel free to remove the hold when all questions have been addressed.

neolit123 commented on cmd/kubeadm/app/util/staticpod/utils.go:
```diff
@@ -321,3 +321,11 @@ func ManifestFilesAreEqual(path1, path2 string) (bool, error) {
 	return bytes.Equal(content1, content2), nil
 }
+
+// getProbeAddress returns a valid probe address
```
this now needs a better explanation of why we are omitting the address here an why it works.

neolit123 replied:
to clarify, we need to explain why the k8s probes work with an empty address.
is this behavior guaranteed long term?

aojea replied:
@thockin is this something you can enlighten us? 
the question was if a probe will work with a 0.0.0.0  address https://github.com/kubernetes/kubernetes/pull/86493/files#r360576622
and we've found that (if my test is correct) it doesn't, however probes with empty hosts works 🤷‍♂

neolit123 replied:
alternatively, you can either a) check the source code of the probes, b) go the alternative route to not feed empty host - instead default to the link-local addresses.

aojea replied:
I've already done a) ;) https://github.com/kubernetes/kubernetes/pull/86505/
let's see how the discussion ends and based on the conclusion we can decide what's the best path for kubeadm

neolit123 replied:
@aojea i suggest we go with b) at this point.

liggitt replied:
> to clarify, we need to explain why the k8s probes work with an empty address.
is this behavior guaranteed long term?

From the API doc: `// Host name to connect to, defaults to the pod IP.` so yes, it is guaranteed long term

neolit123 replied:
ok, let's proceed with merging this PR unless @aojea has reservations.
my minor request here is to document with at least another sentence above `getProbeAddress` why `""` works.

ereslibre replied:
Can't we go one step further and untangle the probe destination address from the bind address? We could always leave the probe to the empty host, so we could simplify this logic even more.

neolit123 replied:
if we are doing that it should be done for the kube-apiserver too.

neolit123 replied:
what if someone is expecting (with external tools) the probe address to be at the bind-address?

@aojea @ereslibre if you decide to go that route, please prefix the release note with "action required: ....".

ereslibre replied:
> what if someone is expecting (with external tools) the probe address to be at the bind-address?

Yeah, I'm fine with the PR as it is right now.

Forcing always the probe host to `""` (as I last proposed) has the drawback that if a user specifies the bind address host to a fixed IP address, then the probe wouldn't work (in this case, where we are using `hostNetwork`), so after a second thought I think it's better leaving the PR as is.

ereslibre commented:
Leaving aside the 0.0.0.0 as destination discussion, why don't we set the empty host for probes? Does this have any drawback?

neolit123 commented:
[this comment](https://github.com/kubernetes/kubernetes/issues/86504#issuecomment-570662492) by thockin makes me suggest to hardcode link-local addresses, instead of using empty host.

ereslibre commented:
> this comment by thockin makes me suggest to hardcode link-local addresses

I don't see why that comment suggests so. Using an empty host should make the kubelet use the pod IP (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/prober/prober.go#L169), so I think leaving this work to the kubelet feels natural, and less cumbersome for us to maintain on the kubeadm side.

Whether probes can use 0.0.0.0 as a destination is a different discussion in my opinion.

neolit123 commented on cmd/kubeadm/app/util/staticpod/utils_test.go:
```diff
@@ -115,12 +129,34 @@ func TestGetControllerManagerProbeAddress(t *testing.T) {
 			expected: "2001:abcd:bcda::1",
 		},
+		{
+			desc: "setting controller manager extra address arg to 0.0.0.0 returns empy",
```
typo `empy`

aojea commented:
/hold cancel
seems that there is consensus 😄

neolit123 commented:
/lgtm
/priority important-longterm

neolit123 commented:
thanks for the updates and discussion.
